hot-reload components referencing k8s secrets (#5234)
dapr/docs: daprdocs/content/en/operations/components/component-secrets.md

    kubectl apply -f ./eventhubs.yaml
    ```

## Updating referenced secrets

### Kubernetes secrets are hot reloaded

> When running in Kubernetes and referencing secrets from the built-in `kubernetes` secret store (that is, when `auth.secretStore` is set to `kubernetes` or left empty), Dapr automatically detects changes to the referenced Kubernetes secrets and reloads the affected components. No restart of the application pod or the Dapr sidecar is required.

Reviewer: ```suggestion
When running in Kubernetes and referencing secrets from the default `kubernetes` secret store (when `auth.secretStore` is set to `kubernetes` or left empty), Dapr automatically detects changes to the referenced Kubernetes secrets and reloads the affected components. No restart of the application pod or the Dapr sidecar is required.
```

Dapr reconciles components every 60 seconds. When the value of a referenced Kubernetes secret changes, [hot reloading]({{% ref "component-updates.md#hot-reloading" %}}) closes the component and re-initializes it using the new secret value.

> For example, with the Redis state store from [Referencing secrets](#referencing-secrets) deployed to Kubernetes, updating the referenced secret:

Reviewer: ```suggestion
For example, if the Redis state store from [Referencing secrets](#referencing-secrets) was deployed to Kubernetes, then updating the referenced secret with the following command causes the state store component to be closed and re-initialized with the new password within 60 seconds. The same happens when the Kubernetes secret is updated by an external secret manager that keeps it in sync with an external source.
```

```bash
kubectl patch secret redis-secret --type merge -p '{"stringData":{"redis-password":"my-new-password"}}'
```

> causes the state store component to be closed and re-initialized with the new password within 60 seconds. The same happens when the Kubernetes secret is updated by an external secret manager that keeps it in sync with an external source.

Reviewer: ```suggestion
```

{{% alert title="Note" color="primary" %}}
Keep in mind the following when relying on this behavior:
- Changes to secret values are picked up by the periodic reconciler, so it can take up to 60 seconds for the new value to be applied.
- The component is unavailable for a short period of time while it is closed and re-initialized.
- This requires [hot reloading]({{% ref "component-updates.md#hot-reloading" %}}) to be enabled, which is the default. If the `HotReload` feature is disabled, secret changes are only picked up when the Dapr sidecar is restarted.

> - Component types that are [excluded from hot reloading]({{% ref "component-updates.md#components-and-subscriptions" %}}) (Actor State Stores and Workflow Backends) do not pick up secret changes and require a restart.

Reviewer: ```suggestion
- Component types that are [excluded from hot reloading]({{% ref "component-updates.md#components-and-subscriptions" %}}) include Actor state stores and Workflow backends, do not pick up secret changes and require a Dapr sidecar restart.
```

{{% /alert %}}

### Other secret stores

> Secrets referenced from any other secret store, for example HashiCorp Vault or Azure Key Vault configured through `auth.secretStore`, are resolved by the Dapr sidecar only once, when the component is initialized. Rotating a secret in these stores does not modify the Component definition, so the change is not detected by hot reloading and the component keeps using the value that was read at initialization time. To apply the new secret value, restart the Dapr sidecar, or trigger a hot reload by applying a change to the component manifest.

Reviewer: ```suggestion
Secrets referenced from any other secret store, for example HashiCorp Vault or Azure Key Vault configured through `auth.secretStore`, are resolved by the Dapr sidecar only once at start-up, when the component is initialized. Rotating a secret in these stores does not modify the Component definition, so the change is not detected by hot reloading and the component keeps using the value that was read at initialization time. To apply the new secret value, restart the Dapr sidecar, or trigger a hot reload by applying a change to the component manifest file.
```

    kubectl apply -f ./eventhubs.yaml
    ```

## Updating referenced secrets

### Kubernetes secrets are hot reloaded

When running in Kubernetes and referencing secrets from the built-in `kubernetes` secret store (that is, when `auth.secretStore` is set to `kubernetes` or left empty), Dapr automatically detects changes to the referenced Kubernetes secrets and reloads the affected components. No restart of the application pod or the Dapr sidecar is required.

> This works because the Dapr operator resolves each `secretKeyRef` against the native Kubernetes secret when serving component definitions to the Dapr sidecars, and each sidecar periodically reconciles its loaded components against the operator every 60 seconds. When the value of a referenced Kubernetes secret changes, the [hot reloading]({{% ref "component-updates.md#hot-reloading" %}}) reconciler detects the changed component, closes it, and re-initializes it using the new secret value.

Reviewer 2: I think there is too much implementation detail here - we just want to say  something along the lines of ...it reconciles every 60s

For example, with the Redis state store from [Referencing secrets](#referencing-secrets) deployed to Kubernetes, updating the referenced secret:

```bash
kubectl patch secret redis-secret --type merge -p '{"stringData":{"redis-password":"my-new-password"}}'
```

causes the state store component to be closed and re-initialized with the new password within 60 seconds.

> This also applies when an external secret manager keeps a native Kubernetes secret in sync. For example, using HashiCorp Vault together with the [Vault Secrets Operator](https://developer.hashicorp.com/vault/docs/platform/k8s/vso):

Reviewer 2: I would drop the vault part - how someone wires to k8s secrets is a user choice - it might cause more confusion
